plat/common/lcpu.c

```diff
@@ -30,14 +37,150 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
+#include <uk/essentials.h>
+#include <uk/arch/atomic.h>
 #include <uk/plat/lcpu.h>
+#include <uk/plat/irq.h>
+#include <uk/plat/time.h>
 #include <uk/plat/common/cpu.h>
+#include <uk/plat/common/lcpu.h>
 #include <uk/plat/common/_time.h>
+#include <uk/list.h>
+#include <uk/print.h>
 
-void ukplat_lcpu_halt(void)
+#include <limits.h>
+#include <errno.h>
+
+/**
+ * Array of LCPUs, one for every CPU in the system.
+ *
+ * TODO: Preferably have a more flexible solution that does not waste memory
+ * for non-present CPUs and does not force us to configure the maximum number
+ * of CPUs beforehand.
+ */
+struct lcpu lcpus[CONFIG_UKPLAT_LCPU_MAXCOUNT];
+
+#ifdef CONFIG_HAVE_SMP
+/**
+ * Number of allocated logical CPUs in the system.
+ * Must be [1, CONFIG_UKPLAT_LCPU_MAXCOUNT) after boot
+ */
+static __u32 lcpu_count = 1;
+
+struct lcpu *lcpu_alloc(__lcpuid id)
+{
+	struct lcpu *lcpu;
+
+	if (lcpu_count == CONFIG_UKPLAT_LCPU_MAXCOUNT)
+		return NULL;
+
+	lcpu = &lcpus[lcpu_count];
+	lcpu->state = LCPU_STATE_OFFLINE;
+	lcpu->id    = id;
+	lcpu->idx   = lcpu_count;
+
+	lcpu_count++;
+
+	return lcpu;
+}
+
+__u32 ukplat_lcpu_count(void)
+{
+	return lcpu_count;
+}
+#else
+#define lcpu_count	(1)
+
+/* Provide weak default implementations for the case that the architecture does
+ * not support multi-processor configurations.
+ */
+__lcpuid __weak lcpu_arch_id(void)
+{
+	return 0;
+}
+
+int __weak lcpu_arch_init(struct lcpu *this_lcpu __unused)
+{
+	return 0;
+}
+#endif /* !CONFIG_HAVE_SMP */
+
+struct lcpu *lcpu_get(__lcpuidx idx)
+{
+	UK_ASSERT(idx < lcpu_count);
+
+	return &lcpus[idx];
+}
+
+struct lcpu *lcpu_get_current(void)
+{
+	return lcpu_get(ukplat_lcpu_idx());
+}
+
+int lcpu_init(struct lcpu *this_lcpu)
+{
+	int rc;
+
+	/*
+	 * NOTE: Do not use anything that might need initialized exception
+	 * traps until after lcpu_arch_init(), as traps might not be
+	 * initialized for this CPU yet!
+	 */
+
+	/* Initialize the bootstrap CPU */
+	if (lcpu_is_bsp(this_lcpu)) {
+		if (unlikely(lcpu_count > 1))
+			return -EPERM;
+
+		this_lcpu->idx   = 0;
+		this_lcpu->id    = lcpu_arch_id();
+		this_lcpu->state = LCPU_STATE_INIT;
```

Comment:
This keeps the bootcore to `INIT` forever, perhaps consider to set instead to `IDLE`?

Reply:
I cannot reproduce this problem. After breaking in one of the APs remote functions and inspecting system state, I get:
```
lcpus[] = {
    state = 0x3, /* BUSY0 */
    idx = 0x0,
    id = 0x0,
    fn = { fn = 0x0, user = 0x0 }
  }, {
    state = 0x4, /* BUSY1 */
    idx = 0x1,
    id = 0x1,
    fn = {fn = 0x401138f0, user = 0x0}
  }, {
    state = 0x4, /* BUSY1 */
    idx = 0x2,
    id = 0x2,
    fn = { fn = 0x401138f0, user = 0x0 }
  }, {
    state = 0x0, /* OFFLINE */
    idx = 0x3,
    id = 0x3,
    s_args = { entry = 0x0, stackp = 0x0
    }
}
```
Also the code is:
```C
	if (lcpu_is_bsp(this_lcpu)) {
		if (unlikely(lcpu_count > 1))
			return -EPERM;

		this_lcpu->idx   = 0;
		this_lcpu->id    = lcpu_arch_id();
		this_lcpu->state = LCPU_STATE_INIT;  <--- set BSP to init
	} else { ... }

	rc = lcpu_arch_init(this_lcpu);              <--- arch init of BSP

        ...

	this_lcpu->state = LCPU_STATE_BUSY0;         <--- sets BSP to BUSY0

	return rc;
```